File: 백준/Gold/2011. 암호코드/암호코드.js

const input = require("fs").readFileSync("/dev/stdin", "utf-8").trim().split("\n");

const code = input[0].trim().split("").map(Number);
const n = code.length;
const MOD = 1000000;

if (n === 0 || code[0] === 0) {
	console.log(0);
	process.exit(0);
}

const dp = Array(n + 1).fill(0);
dp[0] = 1;
dp[1] = 1;

for (let k = 1; k < n; k++) {
	const i = k + 1;
	if (code[k] > 0) {
		dp[i] = (dp[i] + dp[i - 1]) % MOD;
	}
	const twoDigit = code[k - 1] * 10 + code[k];
	if (twoDigit >= 10 && twoDigit <= 26) {
		dp[i] = (dp[i] + dp[i - 2]) % MOD;
	}
}

console.log(dp[n] % MOD);

File: 백준/Gold/2011. 암호코드/암호코드.py
code = list(map(int,input().strip()))
n = len(code)
dp = [0] * (n + 1)
dp[0] = 1
dp[1] = 1

if code[0] == 0:
    print(0)
else:
    for k in range(1, n): # 글자 길이가 2일때 부터 유효함
        i = k + 1 # dp 는 2부터
        if code[k] > 0:
            dp[i] += dp[i - 1]
        if 10 <= code[k] + code[k - 1] * 10 <= 26:
            dp[i] += dp[i - 2] 
    print(dp[n] % 1000000)